firebase_admin/_token_gen.py

```diff
@@ -254,98 +275,134 @@ def __init__(self, **kwargs):
+            raise self._invalid_token_error('Token does not contain required claim "exp"')
+        elif now < issued_at - _CLOCK_SKEW_SECS:
+            raise self._invalid_token_error(
+                'Token used too early, {0} < {1}'.format(now, issued_at))
+        elif now > expires_at + _CLOCK_SKEW_SECS:
+            raise self._expired_token_error(
+                'Token expired, {0} < {1}'.format(expires_at, now), cause=None)
+
+    def _verify_token_signature(self, token, request):
         try:
-            verified_claims = google.oauth2.id_token.verify_token(
-                token,
-                request=request,
-                audience=self.project_id,
-                certs_url=self.cert_url)
-            verified_claims['uid'] = verified_claims['sub']
-            return verified_claims
+            return google.oauth2.id_token.verify_token(
+                token, request=request, audience=self.project_id, certs_url=self.cert_url)
         except google.auth.exceptions.TransportError as error:
             raise CertificateFetchError(str(error), cause=error)
         except ValueError as error:
-            if 'Token expired' in str(error):
-                raise self._expired_token_error(str(error), cause=error)
-            raise self._invalid_token_error(str(error), cause=error)
+            error_factory = (self._expired_token_error if 'Token expired' in str(error) else
+                             self._invalid_token_error)
+            raise error_factory(str(error), cause=error)
 
-    def _decode_unverified(self, token):
-        try:
-            header = jwt.decode_header(token)
-            payload = jwt.decode(token, verify=False)
-            return header, payload
-        except ValueError as error:
-            raise self._invalid_token_error(str(error), cause=error)
+    def verify(self, token, request):
+        """Verifies the signature and data for the provided JWT."""
+        token = token.encode('utf-8') if isinstance(token, six.text_type) else token
+        if not isinstance(token, six.binary_type) or not token:
+            raise ValueError(
+                'Illegal {0} provided: {1}. {0} must be a non-empty '
+                'string.'.format(self.short_name, token))
+
+        if not self.project_id:
+            raise ValueError(
+                'Failed to ascertain project ID from the credential or the environment. Project '
+                'ID is required to call {0}. Initialize the app with a credentials.Certificate '
+                'or set your Firebase project ID as an app option. Alternatively set the '
+                'GOOGLE_CLOUD_PROJECT environment variable.'.format(self.operation))
+
+        header, payload = self._decode_unverified(token)
+
+        self._verify_aud(payload)
+        self._verify_iss(payload)
+        self._verify_sub(payload)
+
+        if not _auth_utils.is_emulator_enabled():
+            self._verify_kid_and_alg(header, payload)
```

Comment:
This changes the order of certain checks (aud and iss verified before header). Not a problem as far as the correctness of the functionality is concerned. But some of our tests might be unhappy with it depending on how they are implemented.

Reply:
Ack'd, the tests as I ran them seemed to be OK so I don't think it's an issue.